Time complexity comments

canadianschism opened the pull request:
So I think I've got the Tc stuff down. It took longer than I expected because I'm still not 100% sure on how the whole thing works. I'm really not sure about the merge sort algo. From what I've been reading online:

- Splitting is O(log n) <- this would be the getMiddleNode part, as it's the splitter using the fast/slow search
- The mergeSort is O(n)
- The sortedMerge is O(n log n)

So, by that logic it would be O(log n), however what I'm seeing is that mergeSort on linked lists are O(n log n) ... So, yeah. If we can get together to discuss this I'd feel better about things.

richieyoum commented:
O(log n) < O(n) < O(n log n). So the time complexity of the algorithm should be O(n log n).

canadianschism commented:
Alright, I've changed things around, along with the answer. As he asked for the answer to be in the main method I've moved it down there, to be O(n log n). I'm not sure if I have to put in another pull request, or if pushes that I'm doing now will automatically be added?

richieyoum commented:
You can simply push it on your branch and it'll reflect in the current PR. Thanks!

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -16,34 +18,34 @@ public SinglyLinkedList(){
      * @return boolean indicating whether list is empty
      */
     public boolean isEmpty(){
+    	// Tc = 1ct
```
```suggestion
    	// Tc = 1c
```
I think you made a typo?

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -16,34 +18,34 @@ public SinglyLinkedList(){
 
     // Access methods
     /**
-     * Returns number of elements in linked list
```
I know it's kind of duplicate info, but javadoc usually includes the descriptions on top of any parameters + return variables. Could you add them back in please? :)

canadianschism replied:
I apologize for this... again, not a programmer, so I didn't know :) I think it should all be back to the way it was.

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -53,17 +55,22 @@ public Node getTailNode(){
      * @param name new bird name
      */
     public void addHeadNode(String name, int count){
+    	// addHeadNode: Tc = 11c
```
shouldn't it be 7c?

line 61 - 2c
line 64 - 1c
line 67 - 1c
line 70 - 1c
line 71 - 0 to 1c
line 74 - 1c

canadianschism replied:
I was thinking that sending something in a method call was 1c ... so it should actually be 12.

Still can't find anything related to method calls and the like, but I remember him saying something like "get the value of the variable" being 1 operation. In my head, calling a new Node(name, count) means that, at the very least, you're getting the value of name and count (2c), and setting it to newNode (1c). I think I'm going to email him about that.

However, at this point, I'm wondering if we shouldn't just be using O(1) for methods whose Tc are constant.

"Comments: in the comments for every method, you have to **specify the time complexity of that method**, and finally for the main method, specify the time complexity of the whole program."

Counting vs. O(1) comes to the same thing, wouldn't it? Your thoughts?

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -72,29 +79,32 @@ public void addHeadNode(String name, int count){
      * @param name new bird name
      */
     public void addTailNode(String name, int count){
+    	// addTailNode: Tc = 12c
```
same comment as the one in addHeadNode

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -72,29 +79,32 @@ public void addHeadNode(String name, int count){
      * Remove the head node
      */
     public void removeHeadNode(){
+    	// removeHeadNode: Tc = 7c
```
I think it's 4c

106 - 1c
108 - 1c
110 - 0 ~ 1c
111 - 0 ~ 1c

canadianschism replied:
Brought down to O(1)

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -1,11 +1,13 @@
 import java.io.IOException;
 
 public class SinglyLinkedList {
+	// 3 operations, constant
```
do we need to count each constant for the assignment? or can we just leave that it's constant, linear ie- O(n), etc...?

canadianschism replied:
I'm doing counts for each method, and then resolving in the main method. I've changed that line to Tc = 3c 

But to be honest, I have zero idea.

canadianschism replied:
This has been taken care of, with each method that is constant being O(1)

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -207,11 +225,12 @@ private Node getMiddleNode(Node tempHead){
-     * @param left a node to be compared with right node
-     * @param right a node to be compared with left node
-     * @return merged node from left and right nodes in ascending order
+     * @param left		: a node to be compared with right node
```
what's up with the indent? :thinking:

canadianschism replied:
I have this thing with stuff lining up on the page. I'll put that back. It was bothering the bejeesus out of me.

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -263,6 +283,7 @@ private Node mergeSort(Node tempHead){
      * Merge sort the current linked list
      */
     public void sortList(){
+    	// sortList: 3c
```
1c

canadianschism replied:
Resolved to O(1) as per Amin

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -271,6 +292,8 @@ public void sortList(){
      * Print the linked list items
      */
     public void printNodeItems(){
+    	// printNodeItems: 3c + 0(n)
```
let's jsut abstract it as constant, so just keep O(n)

canadianschism replied:
Done!

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -341,17 +371,31 @@ public int getCount(){
     }
 
     public static void main(String[] args) throws IOException{
+    	/*	Count of method operations in main method
```
I can't confirm the individual constants, but it is O(n log n). Please update after reviewing my comments above.

canadianschism replied:
Done, I've reviewed, though I thought I'd had O(n log n) in there originally. In any case, reviewed and corrected!

richieyoum commented on assignment1/src/assignment1/SinglyLinkedList.java:
```diff
@@ -36,14 +40,16 @@ public boolean isEmpty(){
 
     /**
-     * Returns list tail node
+     * Returns list tal node
```
tail*

richieyoum commented:
LGTM! Merging now.
